src/com/maddyhome/idea/vim/group/ChangeGroup.java

```diff
@@ -211,6 +215,34 @@ private void runEnterAction(Editor editor, @NotNull DataContext context) {
     }
   }
 
+  /**
+   * Inserts a new line above the caret position
+   */
+  public void insertNewLineAbove(@NotNull Editor editor, @NotNull Caret caret, @NotNull DataContext context) {
```

Comment:
One with a caret given in arguments is for inner use of `ChangeGroup` (it should be private, I guess). It is used by `ChangeLineAction` (help topic `cc`) and `ChangeVisual*Action` (for example, `v_c`). In these actions, the new line is inserted after deletion of some text range and it is decided for each caret whether it should be inserted below (at the end of file) or above (all other cases).

Another one is used by `InsertNewLineAboveAction` (help topic `O`) and `InsertNewLineBelowAction` (help topic `o`). It inserts a new line running an `ENTER` action in the editor. The positive effect of this is that the marks are moved down automatically (if necessary). The negative part is that it is done at once for all of the carets so it is impossible to do this for a concrete caret (so it is impossible to separate cases for each caret as it is done in `changeRange`).

Reply:
The insert new line actions `O` and `o` should insert a new line via `<Enter>` action since they have to insert the correct indent and care about other possible extra actions that `<Enter>` does for various languages in IntelliJ. `O` and `o` should be run on all the carets one by one, not on the main caret. I believe the current implementation in this PR doesn't do it.